Core/src/Geometry/Blueprint.cpp

```diff
@@ -0,0 +1,154 @@
+
+  options.validate();
+
+  if (m_cfg.envelope == ExtentEnvelope::Zero()) {
+    ACTS_WARNING(prefix() << "Root node is configured with zero envelope. This "
+                             "might lead to navigation issues");
+  }
+
+  if (children().size() != 1) {
+    ACTS_ERROR(prefix() << "Root node must have exactly one child");
+    throw std::logic_error("Root node must have exactly one child");
+  }
+
+  auto &child = children().at(0);
+
+  ACTS_DEBUG(prefix() << "Executing building on tree");
+  Volume &topVolume = child.build(options, gctx, logger);
+  const auto &bounds = topVolume.volumeBounds();
+
+  std::stringstream ss;
+  bounds.toStream(ss);
+  ACTS_DEBUG(prefix() << "have top volume: " << ss.str() << "\n"
+                      << topVolume.transform().matrix());
+
+  std::shared_ptr<VolumeBounds> worldBounds;
+
+  if (const auto *cyl = dynamic_cast<const CylinderVolumeBounds *>(&bounds);
+      cyl != nullptr) {
+    using enum CylinderVolumeBounds::BoundValues;
+
+    // Make a copy that we'll modify
+    auto newBounds = std::make_shared<CylinderVolumeBounds>(*cyl);
+
+    const auto &zEnv = m_cfg.envelope[binZ];
+    if (zEnv[0] != zEnv[1]) {
+      ACTS_ERROR(
+          prefix() << "Root node cylinder envelope for z must be symmetric");
+      throw std::logic_error(
+          "Root node cylinder envelope for z must be "
+          "symmetric");
+    }
+
+    const auto &rEnv = m_cfg.envelope[binR];
+
+    newBounds->set({
+        {eHalfLengthZ, newBounds->get(eHalfLengthZ) + zEnv[0]},
+        {eMinR, std::max(0.0, newBounds->get(eMinR) - rEnv[0])},
+        {eMaxR, newBounds->get(eMaxR) + rEnv[1]},
+    });
+
+    worldBounds = std::move(newBounds);
+
+  } else if (const auto *box =
+                 dynamic_cast<const CuboidVolumeBounds *>(&bounds);
+             box != nullptr) {
+    throw std::logic_error{"Not implemented"};
+  } else {
+    throw std::logic_error{"Unsupported volume bounds type"};
+  }
+
+  ACTS_DEBUG(prefix() << "New root volume bounds are: " << *worldBounds);
+
+  auto world = std::make_unique<TrackingVolume>(
+      topVolume.transform(), std::move(worldBounds), "World");
+
+  // @TODO: This needs to become configurable
+  world->setNavigationPolicy(
+      options.defaultNavigationPolicyFactory->build(gctx, *world, logger));
+
+  // Need one-sided portal shell that connects outwards to nullptr
+  SingleCylinderPortalShell worldShell{*world};
+  worldShell.applyToVolume();
+
+  auto &shell = child.connect(options, gctx, logger);
+
+  shell.fill(*world);
+
+  child.finalize(options, gctx, *world, logger);
+
+  std::set<std::string> names;
+
+  world->visitVolumes([&](const auto *volume) {
+    if (names.contains(volume->volumeName())) {
+      ACTS_ERROR(prefix() << "Duplicate volume name: " << volume->volumeName());
+      throw std::logic_error("Duplicate volume name");
+    }
+    names.insert(volume->volumeName());
+  });
```

Comment:
should we also adopt to range-based for here?

Reply:
I'm not completely sure I follow. This traverses the tracking volume tree and invokes the lambda with each volume that's found, similar to `visitSurfaces`. 

The can likely be rewritten as a range based for loop, but that seems non-trivial. Do we need that for this PR you think?

Reply:
Ah this goes through the whole tree? Ok makes sense - I thought if this just handles the first level children we could do something like the transformer range you use in the blueprint node. Maybe something for another PR but does not seem important